src/basicfuns.jl

```diff
@@ -152,9 +152,62 @@ Return `log(1+exp(x))` evaluated carefully for largish `x`.
 
 This is also called the ["softplus"](https://en.wikipedia.org/wiki/Rectifier_(neural_networks))
 transformation, being a smooth approximation to `max(0,x)`. Its inverse is [`logexpm1`](@ref).
+
+See:
+ * Martin Maechler (2012) [“Accurately Computing log(1 − exp(− |a|))”](http://cran.r-project.org/web/packages/Rmpfr/vignettes/log1mexp-note.pdf)
 """
-log1pexp(x::Real) = x < 18.0 ? log1p(exp(x)) : x < 33.3 ? x + exp(-x) : oftype(exp(-x), x)
-log1pexp(x::Float32) = x < 9.0f0 ? log1p(exp(x)) : x < 16.0f0 ? x + exp(-x) : oftype(exp(-x), x)
+function log1pexp(_x::Real)
+    x = float(_x)
+    x0, x1, x2 = _log1pexp_thresholds(x)
+    if x < x0
+        return exp(x)
+    elseif x < x1
+        return log1p(exp(x))
+    elseif x < x2
+        return x + exp(-x)
+    else
+        return x
+    end
+end
+
+#= The precision of BigFloat cannot be computed from the type only.
+Since computing the dynamically thresholds is slow, prefer to use
+a slower but accurate implementation in this case. =#
```

Comment:
This seems contradicting - if the function is slow, we should use a faster alternative but not a slower one? Did you benchmark both approaches?

Reply:
What I meant is that for BigFloat, the approach with the thresholds is slower than this special-cased method. Here are the benchmarks I run.

Method specialized for `BigFloat` without thresholds.

```julia
julia> @benchmark log1pexp(big(x)) setup=(x=30randn();) samples=10^7
BenchmarkTools.Trial: 58177 samples with 7 evaluations.
 Range (min … max):   4.600 μs … 759.421 μs  ┊ GC (min … max): 0.00% … 94.03%
 Time  (median):      8.594 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   11.937 μs ±   8.873 μs  ┊ GC (mean ± σ):  0.41% ±  0.97%

    ▃██▃                                                        
  ▁▄████▆▃▂▁▁▁▁▁▁▁▁▂▄▆▆▆▅▃▂▂▁▁▁▁▁▁▁▁▁▁▁▁▂▃▅▄▄▄▄▆▃▂▂▂▁▁▁▁▁▁▁▁▁▁ ▂
  4.6 μs          Histogram: frequency by time         27.4 μs <

 Memory estimate: 504 bytes, allocs estimate: 11.
```

Method with thresholds.

```
julia> @benchmark log1pexp(big(x)) setup=(x=30randn();) samples=10^7
BenchmarkTools.Trial: 203069 samples with 1 evaluation.
 Range (min … max):  12.555 μs …  4.530 ms  ┊ GC (min … max): 0.00% … 98.47%
 Time  (median):     21.679 μs              ┊ GC (median):    0.00%
 Time  (mean ± σ):   22.388 μs ± 26.404 μs  ┊ GC (mean ± σ):  0.70% ±  0.62%

                    ▄▂█ ▂ ▃                                    
  ▁▁▁▂▅▇▇▆▅▄▃▂▂▁▁▁▁▃█████▆█▆▄▃▂▂▁▁▁▁▁▁▁▂▃▂▃▃▂▃▂▂▃▂▂▁▂▁▁▁▁▁▁▁▁ ▂
  12.6 μs         Histogram: frequency by time        38.5 μs <

 Memory estimate: 1.95 KiB, allocs estimate: 38.
```

 I'll rewrite the comment.

Reply:
Marking resolved.